### PR TITLE
Fix PROTOBUF_GENERATED_MESSAGE_FIELD_OFFSET for Clang 3.7.x and older.

### PR DESCRIPTION
**Pull request by gkrasin**
In Clang older than 3.8, the compiler is not happy about _Pragma("clang diagnostic push").

### CODE REVIEW

**Comment by bazel-io** [comment]
Can one of the admins verify this patch?

File: src/google/protobuf/map_entry.h [hunk 1]
```diff
@@ -248,8 +248,8 @@ class MapEntry : public MapEntryBase {
         entry,
         offsets_,
         has_bits_,
-        GOOGLE_PROTOBUF_GENERATED_MESSAGE_FIELD_OFFSET(MapEntry, entry_lite_._has_bits_),
-        GOOGLE_PROTOBUF_GENERATED_MESSAGE_FIELD_OFFSET(MapEntry, _internal_metadata_),
+        static_cast<int>(GOOGLE_PROTOBUF_GENERATED_MESSAGE_FIELD_OFFSET(MapEntry, entry_lite_._has_bits_)),
```
Comment (xfxyjwf) [inline src/google/protobuf/map_entry.h]: Is this cast necessary?

Comment (gkrasin) [inline src/google/protobuf/map_entry.h]: Yes. Otherwise, Clang complains about an unsigned to signed type implicit cast.

Comment (xfxyjwf) [inline src/google/protobuf/map_entry.h]: @acozzette who recently changed GOOGLE_PROTOBUF_GENERATED_MESSAGE_OFFSET to static cast to uint32. Maybe we should change the corresponding ReflectionSchema field to uint32?

Comment (acozzette) [inline src/google/protobuf/map_entry.h]: That sounds good to me. It sounds like the right way to do it is to use uint32 everywhere consistently.

Comment (gkrasin) [inline src/google/protobuf/map_entry.h]: I have made an attempt to change the type of these two ReflectionSchema fields from int to uint32, but then many more changes have become required. My proposal is to merge this CL to fix the breakage under old Clang and therefore unblock Asad, and do this cleanup separately.

Comment (gkrasin) [inline src/google/protobuf/map_entry.h]: Actually, it's not that bad. Done. Please, take a look.

**Comment by xfxyjwf** [comment]
test this please

**Comment by gkrasin** [comment]
@xfxyjwf was that a magic spell to invoke try bots or a request to test the change locally?

**Comment by xfxyjwf** [comment]
@gkrasin It's a magic spell to trigger the jenkins tests. Sorry, I know it can be a bit confusing to you...

**Comment by xfxyjwf** [comment]
Jenkins reports an unsigned-signed comparison problem:
https://grpc-testing.appspot.com/job/protobuf_pull_request_32/158/console

I guess we need to live with static_cast<int>(GENERATED_MESSAGE_OFFSET()) for now and do a followup cleanup instead.

**Comment by gkrasin** [comment]
I have removed the commit that changed the types of the fields. Please, take another look.

**Comment by xfxyjwf** [comment]
test this please

**Comment by xfxyjwf** [comment]
Closing old issues/prs.

I think we have switched to C++ offsetof now.

**Pull request closed without merging**